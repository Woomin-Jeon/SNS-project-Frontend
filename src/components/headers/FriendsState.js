import React, { useEffect } from 'react';
import func from '../../function';

const callAPI = async (loginState, setLoginState, currentUserState, setCurrentUserState) => {
  const { users } = loginState;

  users.forEach((user, i) => {
    currentUserState.id === users[i].id
      ? setCurrentUserState({
        ...currentUserState,
        friends: users[i].friends,
      })
      : user;
  });
}

function FriendsState({
  specificPost,
  currentUserState,
  setCurrentUserState,
  loginState,
  setLoginState,
}) {
  const { id } = specificPost;
  const { friends } = currentUserState;
  const { users } = loginState;

  useEffect(() => {
    callAPI(loginState, setLoginState, currentUserState, setCurrentUserState);
  }, [loginState]);

  const friendAddButtonClicked = async () => {
    const { userStore } = await func.addFriend(currentUserState.id, id);

    setLoginState({ ...loginState, users: [...userStore] });

    users.forEach((user, i) => {
      currentUserState.id === users[i].id
        ? setCurrentUserState({
          ...currentUserState,
          friends: users[i].friends,
        })
        : user;
    });
  };

  const friendRemoveButtonClicked = async () => {
    const { userStore } = await func.removeFriend(currentUserState.id, id);

    setLoginState({ ...loginState, users: [...userStore] });

    users.forEach((user, i) => {
      currentUserState.id === users[i].id
        ? setCurrentUserState({
          ...currentUserState,
          friends: users[i].friends,
        })
        : user;
    });
  };

  return (
    <>
      <div>
        {friends.includes(id)
          ? (
            <div>
              <span className="others-friend-state">{func.findUserById(users, id).userName} 님과 친구입니다</span>{' '}
              <button
                className="others-friend-already"
                type="button"
                onClick={friendRemoveButtonClicked}
              >
              팔로우 해제
              </button>
            </div>
          ) : (
            <button
              className="others-friend-add"
              type="button"
              onClick={friendAddButtonClicked}
            >
            팔로우
            </button>
          )}
      </div>
    </>
  );
}

export default FriendsState;
