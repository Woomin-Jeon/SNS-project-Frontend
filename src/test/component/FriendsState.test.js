import React from 'react';
import {shallow} from "enzyme";
import FriendsState from "../../components/headers/FriendsState";

describe('FriendsState', () => {
  let currentUserState;
  let setCurrentUserState;
  let specificPost;
  let loginState;
  let setLoginState;
  let p;

  beforeEach(() => {
    specificPost = {
      id: 'Random_ID',
      name: 'Eunjoo'
    };
    currentUserState = {
      id: 'Random_ID_2',
      pw: 'Random_Password',
      userName: 'Woomin',
      friends: []
    };
    loginState = {
      users: [ { id: 'Random_ID_2' }]
    };
    p = {
      time: [2020, 4, 21, 5, 16],
      thumbCount: [],
    };
  });

  it('renders well', () => {
    const component = shallow(
      <FriendsState
        specificPost={specificPost}
        currentUserState={currentUserState}
        setCurrentUserState={setCurrentUserState}
        loginState={loginState}
        setLoginState={setLoginState}
      />
    );

    expect(component).toMatchSnapshot();
    expect(component.find('.others-friend-add').text()).toBe('팔로우');
  });
});
