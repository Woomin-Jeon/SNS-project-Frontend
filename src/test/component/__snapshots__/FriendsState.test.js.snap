// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FriendsState renders well 1`] = `ShallowWrapper {}`;
